completions/completions_v2.py

```diff
@@ -146,11 +133,50 @@ def _run_complete():
                     'next_completion_if_showing': True,
                     'auto_complete_commit_on_tab': True,
                 })
-                
+        EMPTY_COMPLETION_LIST = ([], sublime.INHIBIT_WORD_COMPLETIONS | sublime.INHIBIT_EXPLICIT_COMPLETIONS)
+
+        last_command_insert_snippet = view.command_history(-1)[0] == "insert_snippet"
+
+        if self.is_json_end_line(view, locations[0]):
+            return EMPTY_COMPLETION_LIST
+
+        if prefix.strip() == "" and last_command_insert_snippet:
+            return EMPTY_COMPLETION_LIST
+
+        if view.settings().get("tabnine-disabled", False):
+            return EMPTY_COMPLETION_LIST if prefix.strip() == "" else None
+            
+        if not view.match_selector(locations[0], "source | text"):
+            return EMPTY_COMPLETION_LIST
+            
+        if self.is_query_after_new_line(view, locations[0]):
+            logger.debug("empty character query")
+            return EMPTY_COMPLETION_LIST
```

Comment:
Extract all of this to a method (that returns a boolean, do `if self.method(a,b,c): return EMPTY_COMPLETION_LIST`)